File: iosApp/iosApp/Presentation/Main/CafeManagement/Banner/BannerViewModel.swift
//
//  BannerViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/15.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

final class BannerViewModel: ObservableObject {
    private let cafeId: String?

    private let getHomeBannerManagementUseCase: GetHomeBannerManagementUseCase

    private let deleteHomeBannerUseCase: DeleteHomeBannerUseCase

    private let bannerEventPublisher: BannerEventPublisher

    @Published private(set) var uiState = BannerUiState()

    let event = PassthroughSubject<BannerEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func loadBanners() {
        tasks[.load]?.cancel()
        tasks[.load] = Task {
            do {
                let result = try await getHomeBannerManagementUseCase.invoke(cafeId: cafeId)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let banners = success.data as? [HomeBanner] {
                    let mapped = banners
                        .map { $0.toBannerItem() }
                    uiState.banners = mapped
                    if let pendingDeleteBannerId = uiState.pendingDeleteBannerId,
                       !mapped.contains(where: { $0.id == pendingDeleteBannerId }) {
                        uiState.pendingDeleteBannerId = nil
                    }
                } else {
                    event.send(.showMessage(MessageKey.bannerLoadFailed))
                }
            } catch {
                if Task.isCancelled { return }
                event.send(.showMessage(MessageKey.bannerLoadFailed))
            }
        }
    }

    private func observeBannerEvent() {
        tasks[.bannerEvent]?.cancel()
        tasks[.bannerEvent] = Task {
            do {
                for try await event in asyncSequence(for: bannerEventPublisher.events) {
                    switch event {
                    case is Shared.BannerEvent.Created:
                        self.loadBanners()
                    case let updated as Shared.BannerEvent.Updated:
                        self.patchBanner(updated.banner)
                    case let deleted as Shared.BannerEvent.Deleted:
                        self.removeBanner(id: deleted.banner.id)
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func patchBanner(_ updatedBanner: HomeBanner) {
        let shouldShow: Bool
        if let cafeId = cafeId {
            shouldShow = updatedBanner.cafeId == cafeId
        } else {
            shouldShow = true
        }

        if let index = uiState.banners.firstIndex(where: { $0.id == updatedBanner.id }) {
            if shouldShow {
                uiState.banners[index] = updatedBanner.toBannerItem()
            } else {
                uiState.banners.remove(at: index)
            }
        } else if shouldShow {
            uiState.banners.insert(updatedBanner.toBannerItem(), at: 0)
        }

        if let pendingDeleteBannerId = uiState.pendingDeleteBannerId,
           !uiState.banners.contains(where: { $0.id == pendingDeleteBannerId }) {
            uiState.pendingDeleteBannerId = nil
        }
    }

    func onAction(_ action: BannerAction) {
        switch action {
        case .backTapped:
            event.send(.navigateBack)
        case .selectTab(let tab):
            uiState.selectedTab = tab
        case .createBannerTapped:
            event.send(.navigateToBannerEdit(cafeId: cafeId, bannerId: nil))
        case .editBannerTapped(let id):
            clickEditBanner(id: id)
        case .deleteBannerTapped(let id):
            clickDeleteBanner(id: id)
        case .dismissDeleteBannerDialog:
            uiState.pendingDeleteBannerId = nil
        case .confirmDeleteBanner:
            confirmDeleteBanner()
        }
    }

    private func clickDeleteBanner(id: String) {
        let hasBanner = uiState.banners.contains { $0.id == id }
        if !hasBanner {
            return
        }
        uiState.pendingDeleteBannerId = id
    }

    private func clickEditBanner(id: String) {
        guard let banner = uiState.banners.first(where: { $0.id == id }) else {
            return
        }
        event.send(.navigateToBannerEdit(cafeId: banner.cafeId ?? cafeId, bannerId: banner.id))
    }

    private func confirmDeleteBanner() {
        guard let bannerId = uiState.pendingDeleteBannerId else {
            return
        }

        uiState.pendingDeleteBannerId = nil
        tasks[.delete]?.cancel()
        tasks[.delete] = Task {
            do {
                let result = try await deleteHomeBannerUseCase.invoke(bannerId: bannerId)
                if result is AppResultSuccess<AnyObject> {
                    removeBanner(id: bannerId)
                    event.send(.showMessage(MessageKey.bannerDeleted))
                } else if let failure = result as? AppResultFailure {
                    let _ = failure
                    event.send(.showMessage(MessageKey.bannerDeleteFailed))
                } else {
                    event.send(.showMessage(MessageKey.bannerDeleteFailed))
                }
            } catch {
                if Task.isCancelled { return }
                event.send(.showMessage(MessageKey.bannerDeleteFailed))
            }
        }
    }

    private func removeBanner(id: String) {
        uiState.banners.removeAll { $0.id == id }
        if uiState.pendingDeleteBannerId == id {
            uiState.pendingDeleteBannerId = nil
        }
    }

    init(
        cafeId: String? = nil,
        getHomeBannerManagementUseCase: GetHomeBannerManagementUseCase = KoinInitializerKt.resolveGetHomeBannerManagementUseCase(),
        deleteHomeBannerUseCase: DeleteHomeBannerUseCase = KoinInitializerKt.resolveDeleteHomeBannerUseCase(),
        bannerEventPublisher: BannerEventPublisher = KoinInitializerKt.resolveBannerEventPublisher()
    ) {
        self.cafeId = cafeId
        self.getHomeBannerManagementUseCase = getHomeBannerManagementUseCase
        self.deleteHomeBannerUseCase = deleteHomeBannerUseCase
        self.bannerEventPublisher = bannerEventPublisher
        observeBannerEvent()
        loadBanners()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum MessageKey {
        static let bannerLoadFailed = "banner_info_load_failed"
        static let bannerDeleted = "banner_info_deleted"
        static let bannerDeleteFailed = "banner_info_delete_failed"
    }
}

private enum TaskKey {
    case load
    case delete
    case bannerEvent
}

private extension HomeBanner {
    func toBannerItem() -> BannerItem {
        let tab: BannerTab
        switch statusLabel.uppercased() {
        case "SCHEDULED":
            tab = .scheduled
        case "ENDED", "PAUSED":
            tab = .ended
        default:
            tab = .active
        }
        let statusLabelKey: String
        switch tab {
        case .active:
            statusLabelKey = "banner_status_active"
        case .scheduled:
            statusLabelKey = "banner_status_scheduled"
        case .ended:
            statusLabelKey = "banner_status_ended"
        }
        let icon: String
        switch targetType {
        case .cafeDetail:
            icon = "cup.and.saucer.fill"
        case .eventDetail:
            icon = "calendar.badge.plus"
        case .notice:
            icon = "megaphone.fill"
        default:
            icon = "globe"
        }
        return BannerItem(
            id: id,
            cafeId: cafeId,
            title: title,
            description: subtitle,
            periodDays: displayDays,
            statusLabelKey: statusLabelKey,
            tab: tab,
            accentHex: startColorHex,
            imageIcon: icon,
            imageUrl: imageUrl
        )
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/banner/BannerViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement.banner

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.event.BannerEvent as BannerDomainEvent
import com.hhp227.concafe.domain.event.publisher.BannerEventPublisher
import com.hhp227.concafe.domain.model.BannerLinkTargetType
import com.hhp227.concafe.domain.model.HomeBanner
import com.hhp227.concafe.domain.usecase.DeleteHomeBannerUseCase
import com.hhp227.concafe.domain.usecase.GetHomeFeedUseCase
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class BannerViewModel(
    private val cafeId: String? = null,
    private val getHomeFeedUseCase: GetHomeFeedUseCase,
    private val bannerEventPublisher: BannerEventPublisher,
    private val deleteHomeBannerUseCase: DeleteHomeBannerUseCase
) : ViewModel() {
    private val _uiState = MutableStateFlow(BannerUiState.empty())
    val uiState: StateFlow<BannerUiState> = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<BannerEvent>(replay = 0)
    val event: SharedFlow<BannerEvent> = _event.asSharedFlow()

    private fun loadBanners() {
        viewModelScope.launch {
            when (val result = getHomeFeedUseCase.invoke(popularCastCursor = null, nearbyCafeCursor = null)) {
                is AppResult.Success -> {
                    val targetCafeId = cafeId
                    val mapped = result.data.banners
                        .asSequence()
                        .filter { banner -> targetCafeId.isNullOrBlank() || banner.cafeId == targetCafeId }
                        .map { banner -> banner.toBannerItem() }
                        .toList()
                    _uiState.update { state ->
                        state.copy(
                            banners = mapped,
                            pendingDeleteBannerId = state.pendingDeleteBannerId?.takeIf { pendingId ->
                                mapped.any { it.id == pendingId }
                            }
                        )
                    }
                }
                is AppResult.Failure -> {
                    _event.emit(BannerEvent.ShowMessage("배너 목록을 불러오지 못했습니다."))
                }
            }
        }
    }

    private fun observeBannerEvent() {
        viewModelScope.launch {
            bannerEventPublisher.events.collectLatest { event ->
                when (event) {
                    is BannerDomainEvent.Created -> loadBanners()
                    is BannerDomainEvent.Updated -> patchBanner(event.banner)
                    is BannerDomainEvent.Deleted -> removeBanner(event.banner.id)
                }
            }
        }
    }

    private fun patchBanner(updatedBanner: HomeBanner) {
        _uiState.update { state ->
            val targetCafeId = cafeId
            val shouldShow = targetCafeId.isNullOrBlank() || updatedBanner.cafeId == targetCafeId
            val existingIndex = state.banners.indexOfFirst { it.id == updatedBanner.id }

            when {
                shouldShow && existingIndex >= 0 -> {
                    val patched = state.banners.toMutableList()
                    patched[existingIndex] = updatedBanner.toBannerItem()
                    state.copy(banners = patched)
                }
                shouldShow && existingIndex < 0 -> {
                    state.copy(banners = listOf(updatedBanner.toBannerItem()) + state.banners)
                }
                !shouldShow && existingIndex >= 0 -> {
                    val filtered = state.banners.filterNot { it.id == updatedBanner.id }
                    state.copy(
                        banners = filtered,
                        pendingDeleteBannerId = state.pendingDeleteBannerId?.takeIf { pendingId ->
                            filtered.any { it.id == pendingId }
                        }
                    )
                }
                else -> state
            }
        }
    }

    private fun removeBanner(bannerId: String) {
        _uiState.update { state ->
            state.copy(
                banners = state.banners.filterNot { it.id == bannerId },
                pendingDeleteBannerId = state.pendingDeleteBannerId?.takeUnless { it == bannerId }
            )
        }
    }

    private fun clickDeleteBanner(bannerId: String) {
        val hasBanner = _uiState.value.banners.any { it.id == bannerId }
        if (!hasBanner) {
            return
        }
        _uiState.update { it.copy(pendingDeleteBannerId = bannerId) }
    }

    private fun dismissDeleteBannerDialog() {
        _uiState.update { it.copy(pendingDeleteBannerId = null) }
    }

    private fun confirmDeleteBanner() {
        val bannerId = _uiState.value.pendingDeleteBannerId
        if (bannerId == null) {
            return
        }
        viewModelScope.launch {
            when (val result = deleteHomeBannerUseCase.invoke(bannerId)) {
                is AppResult.Success -> {
                    removeBanner(bannerId)
                    _event.emit(BannerEvent.ShowMessage("배너를 삭제했습니다."))
                }
                is AppResult.Failure -> {
                    _uiState.update { it.copy(pendingDeleteBannerId = null) }
                    _event.emit(BannerEvent.ShowMessage(result.error.toString()))
                }
            }
        }
    }

    fun onAction(action: BannerAction) {
        when (action) {
            BannerAction.ClickBack -> emitEvent(BannerEvent.NavigateBack)
            BannerAction.ClickCreateBanner -> emitEvent(BannerEvent.NavigateToBannerEdit(cafeId = cafeId))
            BannerAction.ConfirmDeleteBanner -> confirmDeleteBanner()
            BannerAction.DismissDeleteBannerDialog -> dismissDeleteBannerDialog()
            is BannerAction.ClickDeleteBanner -> clickDeleteBanner(action.bannerId)
            is BannerAction.ClickEditBanner -> clickEditBanner(action.bannerId)
            is BannerAction.SelectTab -> _uiState.update { it.copy(selectedTab = action.tab) }
        }
    }

    private fun clickEditBanner(bannerId: String) {
        val banner = _uiState.value.banners.firstOrNull { it.id == bannerId } ?: return
        emitEvent(
            BannerEvent.NavigateToBannerEdit(
                cafeId = banner.cafeId ?: cafeId,
                bannerId = banner.id
            )
        )
    }

    private fun emitEvent(event: BannerEvent) {
        viewModelScope.launch {
            _event.emit(event)
        }
    }

    init {
        observeBannerEvent()
        loadBanners()
    }
}

private fun HomeBanner.toBannerItem(): BannerItem {
    val tab = when (statusLabel.uppercase()) {
        "SCHEDULED" -> BannerTab.SCHEDULED
        "ENDED", "PAUSED" -> BannerTab.ENDED
        else -> BannerTab.ACTIVE
    }
    val status = when (tab) {
        BannerTab.ACTIVE -> "진행 중"
        BannerTab.SCHEDULED -> "예약"
        BannerTab.ENDED -> "종료"
    }
    val icon = when (targetType) {
        BannerLinkTargetType.CAFE_DETAIL -> "local_cafe"
        BannerLinkTargetType.EVENT_DETAIL -> "event"
        BannerLinkTargetType.NOTICE -> "campaign"
        BannerLinkTargetType.EXTERNAL_LINK -> "language"
    }
    return BannerItem(
        id = id,
        cafeId = cafeId,
        title = title,
        description = subtitle,
        periodText = "노출 ${displayDays}일",
        statusLabel = status,
        tab = tab,
        accentColorHex = startColorHex,
        imageIcon = icon
    )
}
